src/netops.c

```diff
@@ -280,3 +280,89 @@ int gitno_extract_url_parts(
 
 	return 0;
 }
+
+int gitno_extract_url_data_parts(
+	char **scheme,
+	char **host,
+	char **port,
+	char **path,
+	char **username,
+	char **password,
+	char **query,
+	const char *url)
+{
+	struct http_parser_url u = { 0 };
+	const char *_userinfo;
+
+	if(http_parser_parse_url(url, strlen(url), false, &u)) {
+		giterr_set(GITERR_NET, "Malformed URL '%s'", url);
+		return GIT_EINVALIDSPEC;
+	}
+
+#define GITNO_COPY_URL_PART(dest, field) \
```

Comment:
I'm not entirely sure I understand what you're saying here. My assumption is that you're suggesting I wrap the if statement in a do...while(0) to prevent someone from accidentally deleting the braces thus causing it to always return GIT_EINVALIDSPEC. It seems like a far more pragmatic approach would be to require all control flow statements to use braces instead of sprinkling do...while(0)'s around everywhere - this is a well established requirement for preventing the aforementioned bug when writing safety critical software.